### PR TITLE
Only setdefault for signatures if device has key_json

### PR DESCRIPTION
**Pull request by anoadragon453**
A possible fix for https://github.com/matrix-org/synapse/issues/7163

I'm not entirely sure what this code is doing, but I know it's incorrect that we're trying to pull a key out of a dictionary all the time, when we're only setting that key sometimes (when `key_json` is truthy).

~~Stopped the exception on Cadair's server (which is expected), however I'm not convinced that this isn't just fixing a symptom of another issue.~~ (@uhoreg seems to think this does the right thing). Considering this code was introduced in 1.11.0 with https://github.com/matrix-org/synapse/pull/6844, but has only been reported with 1.12.0.

### CODE REVIEW

**Comment by anoadragon453** [comment]
I wonder if https://github.com/matrix-org/synapse/pull/6956 is related.

**Comment by anoadragon453** [comment]
> I wonder if #6956 is related.

Apparently it is not.

**Comment by uhoreg** [comment]
> Considering this code was introduced in 1.11.0 with #6844, but has only been reported with 1.12.0.

It could be due to a change on the client side.  This could be triggered by a client uploading signatures of the device keys, but failing to upload the keys themselves.

File: changelog.d/7177.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Mitigation for a bug in `_get_e2e_device_keys_txn` which broke federation for some servers.
```
Comment (richvdh) [inline changelog.d/7177.bugfix]: only a mitigation?

also I don't really think the reader cares which function it's in.  How about:

> Fix a bug which could cause outbound federation traffic to stop working if a client uploaded an incorrect e2e device signature.

**Pull request merged**